File: Array/Prefix Sum or Sum based approach/Flip to Maximize 1s/kadane_algo.java
import java.util.*

// Time complexity :- O(n)
// Space complexity :- O(1)

// Approach :- 
// convert 0's to +1 andd 1's to -1 and apply kadane algo on the array 
// and record the range for [left, right] for the maximum answer 
// update the final sum of the range 

// Links :- https://www.geeksforgeeks.org/problems/flip-bits0240/1

class Solution {
    int maxOnes(int[] arr) {
        int n = arr.length;
        int left = 0, right = -1;
        int rangeLeft = -1, rangeRight = -1;
        int currSum = 0, totalSum = 0, maxSum = 0;
        
        
        // kadanes algorithm 
        for(int i=0; i<n; ++i){
            // overall sum of array 
            totalSum += arr[i];
            
            // sum of the current element 
            if(arr[i] == 1){
                currSum -= 1;
            }else{
                currSum += 1;
            }
            
            // update the max sum 
            if(currSum > maxSum){
                maxSum = currSum;
                right = i;
                
                // update the range 
                rangeLeft = left;
                rangeRight = right;
            }
            // curr sum becomes -ve 
            if(currSum < 0){
                // update the sum and left details 
                currSum = 0;
                left = i+1;
            }
        }
        // all 1's value 
        if(rangeLeft == -1)
            return totalSum;
        
        // upate the final sum for the range 
        for(int i=rangeLeft; i<=rangeRight; ++i){
            if(arr[i] == 1){
                totalSum -= 1;
            }else{
                totalSum += 1;
            }
        }
        return totalSum;
    }
};

File: Dynamic Programming/Kadane Algo/Max Absolute Diff of Two Subarrays/kadane_algo.cpp
#include<bits/stdc++.h>
using namespace std;

// Time complextiy :- O(n)
// Space complexity :- O(n)

// Approach :-
// inorder to find the absolute difference b/w two subarray, we can split the subarray into two halves at every index 
// for first subarray we'll find the maximum subarray sum and second subarray we'll find the minimum subarray sum or vice versa 
// will find the absolute difference b/w them and maintain the maximum among all of them 
// we can use kadane algorithm to find the maximum and minimum subarray sum 


// Link :- https://www.geeksforgeeks.org/problems/max-absolute-difference4114/1

class Solution {
  public:
    int maxDiffSubArrays(vector<int>& arr) {
        int n = arr.size();
        int maxSum = INT_MIN, sum1 = 0;
        int minSum = INT_MAX, sum2 = 0;
        vector<int> rightMaxSum(n), rightMinSum(n);
        
        // right to left traversal 
        for(int i=n-1; i>=0; --i){
            // kadane algo to find maximum sum 
            sum1 += arr[i];
            maxSum = max(maxSum, sum1);
            if(sum1 < 0){
                sum1 = 0;
            }
            rightMaxSum[i] = maxSum; 
            
            // kadane algo to find minimum sum 
            sum2 += arr[i];
            minSum = min(maxSum, sum2);
            if(sum2 > 0){
                sum2 = 0;
            }
            rightMinSum[i] = minSum;
            
        }
        
        maxSum = INT_MIN, sum1 =0;
        minSum = INT_MAX, sum2 = 0;
        
        int res = 0;
        
        // left to right traversal 
        for(int i=0; i<n; ++i){
            // kadane algo to find maximum sum 
            sum1 += arr[i];
            maxSum = max(maxSum, sum1);
            if(sum1 < 0){
                sum1 = 0;
            }
            int diff = INT_MIN;
            if(i+1 < n){
                diff = max(diff, maxSum - rightMinSum[i+1]);   
            }
            
            // kadane algo to find minimum sum 
            sum2 += arr[i];
            minSum = min(maxSum, sum2);
            if(sum2 > 0){
                sum2 = 0;
            }
            if(i+1 < n){
                diff = max(diff, rightMaxSum[i+1] - minSum);
            }
            res = max(res, diff);
        }
        return res;
    }
};